### PR TITLE
Paging operation name

### PR DESCRIPTION
**Pull request by MiYanni**
Fixes: https://github.com/Azure/autorest.csharp/issues/397

This adds in supporting `operationName`. There is only 1 test that I can make pass right now with this change. The other tests need https://github.com/Azure/autorest.csharp/issues/411.

### CODE REVIEW

File: src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs [hunk 1]
```diff
@@ -34,31 +34,49 @@ public static Client BuildClient(OperationGroup operationGroup)
...
-                    methods.Add(method);
-                    var pageable = operation.Extensions.GetValue<IDictionary<object, object>>("x-ms-pageable");
-                    if (pageable != null)
+                    operationMethods.Add(operation.Language.Default.Name, (operation, method));
+                }
+            }
+
+            List<Method> nextPageMethods = new List<Method>();
+            List<Paging> pagingMethods = new List<Paging>();
+            foreach ((string _, (Operation operation, Method method)) in operationMethods)
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Lets add a private struct for `(Operation operation, Method method)`. Right now it makes code very verbose.

Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: operationMethods.Values?

File: src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs [hunk 2]
```diff
@@ -34,31 +34,49 @@ public static Client BuildClient(OperationGroup operationGroup)
...
+            {
+                IDictionary<object, object>? pageable = operation.Extensions.GetValue<IDictionary<object, object>>("x-ms-pageable");
+                if (pageable != null)
+                {
+                    //TODO: Assuming operationName is in this operation group: https://github.com/Azure/autorest.modelerfour/issues/85
+                    string? operationName = pageable.GetValue<string>("operationName")?.Split('_').Last();
+                    (Operation NextOperation, Method NextMethod)? next =
+                        operationName != null && operationMethods.TryGetValue(operationName,
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Should we throw if the `operationName` is not found?

Comment (MiYanni) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: I really, really want M4 to handle stuff like this. IMO, at our point in the process, we shouldn't need to validate the data we received. But, this is a weird case. Do we error for anything else related to the data right now?

Comment (pakrym) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: > I really, really want M4 to handle stuff like this.

Me too.

I'm just trying to figure out what would be the most confusing option for customers getting a new method generated or an error.

> Do we error for anything else related to the data right now?

We are, often not cleanly, but a parameter that's defined in path but not in parameter list would result in error.

Comment (MiYanni) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Thinking about it a bit more, we should probably **not** go with a design that 'massages' issues away. I do have insight from the PowerShell generator around errors. I only had about 3-4 errors that could happen. Whenever a customer hit one, they wouldn't look into the error. They either created an issue in GitHub or messaged us directly about it. Part of the problem was where the error occurred in the code, we didn't have all the data necessary to make a detailed, actionable error message. In this situation, I can at least make a message that has enough context for them to investigate. So, I'll add that now.

Comment (MiYanni) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Error added. Also created: https://github.com/Azure/autorest.testserver/issues/129

File: src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs (hunk 1 above)
Comment (MiYanni) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Done.

Comment (MiYanni) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Done.

File: src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs [hunk 3]
```diff
@@ -34,34 +34,72 @@ public static Client BuildClient(OperationGroup operationGroup)
...
-                        methods.Add(nextPageMethod);
-                        //TODO: This is a hack since we don't have the model information at this point
-                        var schemaForPaging = ((method.Response.ResponseBody as ObjectResponseBody)?.Type as SchemaTypeReference)?.Schema as ObjectSchema;
-                        Paging pagingMethod = GetClientMethodPaging(method, nextPageMethod, pageable, schemaForPaging);
-                        pagingMethods.Add(pagingMethod);
+                        if (!processedMethods.TryGetValue(operationName, out OperationMethod nextOperationMethod))
+                        {
+                            string processedOperationName = processed.Operation.Language.Default.Name;
+                            throw new Exception(
+                                $"The x-ms-pageable operationName \"{operationGroup.Key}_{operationName}\" for operation {operationGroup.Key}_{processedOperationName} was not found.");
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Use the original value of `operationName` extension here, no need to split and then re-join.

Comment (MiYanni) [reply src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs]: Done.

**Pull request merged**